### PR TITLE
Changing the bytes/CharIterator to the ByteStream

### LINKED ISSUE #384: use gosub_shared::ByteStream instead of gosub_shared:Byte
The gosub_shared::ByteStream works a bit better (?) than the byte structure. The CSS3 uses the ByteStream, but the HTML5 parser still uses Bytes (and maybe some other places too)..  We should move all streams to the ByteStream, and later on improve this stream reader (for instance, to work better with encoding)

### PR DESCRIPTION
**Pull request by jaytaph**
This commit mostly removes all the bytes::CharIterator code and replaces it with the ByteStream. This is a more clean setup of reading a stream of characters taken from a stream of bytes. Now both the HTML5 and CSS3 parser will work with the same bytestream system which makes things a little more clear in the end.

As an additional bonus: attributes are now checked during testing, and all tests seem to pass!

Fixes #384

### CODE REVIEW

File: crates/gosub_css3/src/tokenizer.rs [hunk 1]
```diff
@@ -906,37 +896,29 @@ impl<'stream> Tokenizer<'stream> {
...
+        self.stream.offset()
     }
 
-    pub fn slice(&self, start: usize, end: usize) -> String {
+    pub fn slice(&mut self, start: usize, end: usize) -> String {
+        let old_pos = self.stream.offset();
+        self.stream.seek(start);
+
+        // todo: this is not efficient
         let mut s = String::new();
```
Comment (Sharktheone) [inline crates/gosub_css3/src/tokenizer.rs:907]: You could use `String::with_capacity(end - start)` (or maybe even only do the calculation once (or maybe the compiler does that?))

File: crates/gosub_html5/src/tokenizer.rs [hunk 2]
```diff
@@ -337,25 +323,28 @@ impl<'stream> Tokenizer<'stream> {
...
+                    match c {
+                        Ch('/') => {
+                            self.temporary_buffer.clear();
+                            self.state = State::RCDATAEndTagOpen;
+                        }
+                        _ => {
+                            self.consume('<');
+                            self.stream_prev();
+                            self.state = State::RCDATA;
+                        }
```
Comment (Sharktheone) [inline crates/gosub_html5/src/tokenizer.rs:347]: I guess, this should be a `if let` match pattern again, in my clippy PR, this actually will be an warning.

File: crates/gosub_html5/src/tokenizer.rs [hunk 3]
```diff
@@ -419,32 +408,32 @@ impl<'stream> Tokenizer<'stream> {
...
+                    match c {
+                        Ch('/') => {
+                            self.temporary_buffer.clear();
+                            self.state = State::RAWTEXTEndTagOpen;
+                        }
+                        _ => {
+                            self.consume('<');
+                            self.stream_prev();
+                            self.state = State::RAWTEXT;
+                        }
```
Comment (Sharktheone) [inline crates/gosub_html5/src/tokenizer.rs:436]: same here

File: crates/gosub_html5/src/tokenizer.rs [hunk 4]
```diff
@@ -1899,66 +1952,77 @@ impl<'stream> Tokenizer<'stream> {
...
-                        self.state = State::CDATASection;
+                    match c {
+                        Ch(']') => {
+                            self.state = State::CDATASectionEnd;
+                        }
+                        _ => {
+                            self.consume(']');
+                            self.stream_prev();
+                            self.state = State::CDATASection;
+                        }
```
Comment (Sharktheone) [inline crates/gosub_html5/src/tokenizer.rs:2009]: should also be a if let

**Review by Sharktheone — CHANGES_REQUESTED** [review]
Only some minor things, but the rest looks good.

File: crates/gosub_html5/tests/data/html5lib-tests/tokenizer/contentModelFlags.test [hunk 5]
```diff
@@ -1,52 +1,5 @@
...
-"output":[["Character", "foo</xmp"]]},
-
-{"description":"End tag closing RCDATA or RAWTEXT (ending with slash)",
-"initialStates":["RCDATA state", "RAWTEXT state"],
-"lastStartTag":"xmp",
-"input":"foo</xmp/",
-"output":[["Character", "foo"]],
-"errors":[
-    { "code": "eof-in-tag", "line": 1, "col": 10 }
-]},
```
Comment (emwalker) [inline crates/gosub_html5/tests/data/html5lib-tests/tokenizer/contentModelFlags.test:49]: What happened to these test files?

**Review by emwalker — APPROVED** [review]
Seems good to me.  A pity to see the test files go away.

File: crates/gosub_html5/tests/data/html5lib-tests/tokenizer/contentModelFlags.test (hunk 5 above)
Comment (jaytaph) [reply crates/gosub_html5/tests/data/html5lib-tests/tokenizer/contentModelFlags.test:49]: that should not be... I'll check what is going on..

Comment (jaytaph) [reply crates/gosub_html5/tests/data/html5lib-tests/tokenizer/contentModelFlags.test]: @emwalker I've added them back.. I accidentally committed edits in the tests and failed to restore them first

File: crates/gosub_css3/src/tokenizer.rs (hunk 1 above)
Comment (jaytaph) [reply crates/gosub_css3/src/tokenizer.rs:907]: fixed

File: crates/gosub_html5/src/tokenizer.rs (hunk 2 above)
Comment (jaytaph) [reply crates/gosub_html5/src/tokenizer.rs:347]: fixed

File: crates/gosub_html5/src/tokenizer.rs (hunk 4 above)
Comment (jaytaph) [reply crates/gosub_html5/src/tokenizer.rs:2009]: fixed

File: crates/gosub_html5/src/tokenizer.rs (hunk 3 above)
Comment (jaytaph) [reply crates/gosub_html5/src/tokenizer.rs]: fixed

**Review by Sharktheone — APPROVED** [review]

**Pull request merged**